[TrapFocus] Add documentation (#22062)
mui-org/material-ui: docs/src/pages/components/trap-focus/trap-focus.md

---
title: Trap Focus React component
components: Unstable_TrapFocus
---

# Trap Focus

<p class="description">Trap focus within a DOM node.</p>

> `TrapFocus` is a utility component that manage focus for its descendants.

Reviewer: ```suggestion
`TrapFocus` is a utility component that manages focus for its descendants.
```

> This is useful when implementing overlays like modal dialogs, which should not allow focus to escape while open.

Reviewer: ```suggestion
This is useful when implementing overlays such as modal dialogs, which should not allow focus to escape while open.
```

> This is useful when implementing overlays like modal dialogs, which should not allow focus to escape them while open.

Reviewer: ```suggestion
This is useful when implementing overlays like modal dialogs, which should not allow focus to escape while open.
```

> When `open={true}` the trap enables, pressing <kbd>Tab</kbd> or <kbd>Shift+Tab</kbd> will circle focus within the inner focusable elements of the component.

Reviewer 2: I was tempted to use `<kbd>` somewhere else but I wasn't sure how screen readers handle them?

Author: Nothing special in VoiceOver

<img width="965" alt="Capture d’écran 2020-08-04 à 23 43 50" src="https://user-images.githubusercontent.com/3165635/89348158-61197a00-d6ac-11ea-9258-647797434d82.png">

Reviewer 2: Tested it in JAWS and NVDA, seemed fine.

> When `open={true}` the trap enables, pressing <kbd>Tab</kbd> or <kbd>Shift+Tab</kbd> will circle focus within the inner focusable elements of the component.

Reviewer: ```suggestion
When `open={true}` the trap is enabled, and pressing <kbd>Tab</kbd> or <kbd>Shift+Tab</kbd> will rotate focus within the inner focusable elements of the component.
```

- 📦 [1.5 kB gzipped](https://material-ui.com/size-snapshot).
- ⚛️ Support portals

## Example

{{"demo": "pages/components/trap-focus/BasicTrapFocus.js"}}

## Disable enforce focus

Clicks within the focus trap behave normally; but clicks outside the focus trap are blocked.

> You can disable the behavior with the `disableEnforceFocus` prop.

Reviewer: ```suggestion
You can disable this behavior with the `disableEnforceFocus` prop.
```

{{"demo": "pages/components/trap-focus/DisableEnforceFocus.js"}}

## Disable restore focus

> The component restores the focus back to the previously focused element when it deactivates, for example, back to a button which opened a dialog.

Reviewer: ```suggestion
The component restores the focus back to the previously focused element when it deactivates, for example back to the button which opened a dialog.
```

You can disable this behavior with the `disableRestoreFocus` prop.

{{"demo": "pages/components/trap-focus/DisableRestoreFocus.js"}}

## Lazy activation

> By default, the component moves the focus to its descendants as soon as it opens, `open={true}`.

Reviewer: ```suggestion
By default, the component moves the focus to its descendants as soon as it opens: `open={true}`.
```

> You can disale this behavior, and make it lazy, with the `disableAutoFocus` prop.

Reviewer: ```suggestion
You can disable this behavior and make it lazy with the `disableAutoFocus` prop.
```

{{"demo": "pages/components/trap-focus/LazyTrapFocus.js"}}

## Portal

> The following demo uses [`Portal`](/components/portal/) to render a part of the trap focus into a new "subtree" outside of current DOM hierarchy.

Reviewer: ```suggestion
The following demo uses [`Portal`](/components/portal/) to render a part of the trap focus into a new "subtree" outside of the current DOM hierarchy.
```

> The following demo uses [`Portal`](/components/portal/) to render a part of the trap focus into a new "subtree" outside of the current DOM hierarchy.

Reviewer: ```suggestion
The following demo uses the [`Portal`](/components/portal/) component to render a subset of the trap focus children into a new "subtree" outside of the current DOM hierarchy, so that they no longer form part of the focus loop.
```
Unsure about terminology for this one...

---
title: Trap Focus React component
components: Unstable_TrapFocus
---

# Trap Focus

<p class="description">Trap focus within a DOM node.</p>

> ⚠️ The component is experimental and unstable.

> `TrapFocus` is a utility component that manages focus for its descendants.

Reviewer 3: Is there any function in calling it a "utility component"? It's a component. I think that's sufficient.

This is useful when implementing overlays such as modal dialogs, which should not allow focus to escape while open.

> When `open={true}` the trap is enabled, and pressing <kbd>Tab</kbd> or <kbd>Shift+Tab</kbd> will rotate focus within the inner focusable elements of the component.

Reviewer 3: ```suggestion
When `open={true}` the trap is enabled, and pressing <kbd>Tab</kbd> or <kbd><kbd>Shift</kbd>+<kbd>Tab</kbd></kbd> will rotate focus within the inner focusable elements of the component.
```
See [Representing keystrokes within an input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/kbd#Representing_keystrokes_within_an_input)

- 📦 [1.5 kB gzipped](https://material-ui.com/size-snapshot).
- ⚛️ Support portals

## Example

{{"demo": "pages/components/trap-focus/BasicTrapFocus.js"}}

## Disable enforce focus

Clicks within the focus trap behave normally; but clicks outside the focus trap are blocked.

You can disable this behavior with the `disableEnforceFocus` prop.

{{"demo": "pages/components/trap-focus/DisableEnforceFocus.js"}}

## Disable restore focus

The component restores the focus back to the previously focused element when it deactivates, for example back to the button which opened a dialog.

> You can disable this behavior with the `disableRestoreFocus` prop.

Reviewer 3: We should either explain a concrete use case or not document it. Why would I want to disable it? What are the drawbacks of this prop? etc.

Author: I have removed it

{{"demo": "pages/components/trap-focus/DisableRestoreFocus.js"}}

## Lazy activation

By default, the component moves the focus to its descendants as soon as it opens: `open={true}`.

> You can disable this behavior and make it lazy with the `disableAutoFocus` prop.

Reviewer 3: What part of the demo showcases the lazy part? As far as I can tell it only disables the autofocus. I can see no additional or lazy autofocus.

Author: I have added another sentence to better explain the behavior

Reviewer 3: It's still not clear where the lazy part is in the demo? Do you mean that the activation only triggers once the focus is inside? I think we can describe this better and add an element outside of the trap that is focuable

Reviewer 3: Ah, you added another element to illustrate it.
